util/src/timestamp.rs

```
@@ -1,22 +1,120 @@
+    ntp::request(addr)
+        .map(|p| {
+            let ts = local_time(p.receive_timestamp);
+
+            (ts.timestamp(), ts.timestamp_subsec_nanos())
+        })
+        .ok()
+}
+
+/// Function to get timestamp from system as UTC Unix timestamp, seconds since Unix epoch
+pub fn get_timestamp() -> (i64) {
+    get_timestamp_nanos().0
 }
 
 /// Function to get timestamp from system as UTC Unix timestamp, seconds and nanoseconds since Unix epoch
 pub fn get_timestamp_nanos() -> (i64, u32) {
     // Get UTC current datetime
     let utc: DateTime<Utc> = Utc::now();
 
-    // Return number of non-leap seconds since Unix epoch and the number of nanoseconds since the last second boundary
-    (utc.timestamp(), utc.timestamp_subsec_nanos())
+    let utc_secs = utc.timestamp();
+    let utc_subsec_nanos = utc.timestamp_subsec_nanos();
+
+    let mut old_ts = get_global_timestamp();
+
+    if utc_secs - old_ts.last_ntp.0 > old_ts.update_period {
+        let ntp = get_timestamp_ntp(old_ts.server.as_str())
+            // In case of error, use timestamp from system
+            .unwrap_or((utc_secs, utc_subsec_nanos));
+
+        if let Some(diff) = duration_between_timestamps((utc_secs, utc_subsec_nanos), ntp) {
+            old_ts.last_ntp = ntp;
+            old_ts.diff = diff;
+            old_ts.bigger = true;
+        } else {
+            let diff = duration_between_timestamps(ntp, (utc_secs, utc_subsec_nanos)).unwrap();
+            old_ts.last_ntp = ntp;
+            old_ts.diff = diff;
+            old_ts.bigger = false;
+        }
+
+        set_global_timestamp(old_ts.clone());
+    }
+
+    let utc_dur = Duration::new(utc_secs as u64, utc_subsec_nanos);
+    let result = if old_ts.bigger {
+        utc_dur.checked_add(old_ts.diff)
+    } else {
+        utc_dur.checked_sub(old_ts.diff)
+    };
+
+    match result {
+        Some(x) => (x.as_secs() as i64, x.subsec_nanos()),
+        None => (utc_secs, utc_subsec_nanos),
```

This `None` means that there has been an overflow while calculating the timestamp. I think it is better to panic than to return a potentially invalid timestamp.